Tests for scripts in io/fits/scripts

mohanagr opened the pull request:
Implements test for script `fitsheader` as of now.

### Reference:
Issue #5799

mohanagr commented:
Table formatting is yet to be tested. Working on that.

MSeifert04 commented:
I think some unrelated changes (commits) made it into this PR. Could you clean them up (remove them) so it only contains "new" changes?

mohanagr commented:
Oh. My bad. Sorry.

MSeifert04 commented on astropy/io/fits/tests/test_fitsheader.py:
```diff
@@ -0,0 +1,159 @@
+
+    @pytest.mark.parametrize('test_file', [
+        ('test0.fits'),
+        pytest.mark.xfail(('random.fits'), reason = 'bad file')
```
Why create a test just to `xfail` it? Or am I misunderstand this line?

mohanagr replied:
There are two tests, one is supposed to pass i.e. `test0.fits` does exist and other `xfails`.

MSeifert04 replied:
But a test that should fail shouldn't be `xfail` but test it with `pytest.raises`?

bsipocz replied:
I agree with @MSeifert04. ``xfail`` shouldn't be used for this case, if a failure is expected and normal behaviour then it should be caught with ``pytest.raises``.

mohanagr replied:
@MSeifert04 @bsipocz All right. What are the use cases for `xfail` then?

MSeifert04 replied:
- `xfail` - should work but currently doesn't because of some unfixed bug (either in-package or upstream)
- `skip` - doesn't work if dependency is too old or if dependency is not installed
- `pytest.raises` - should fail, make sure it fails with the right exception (and exception message).

That's not official, that's just how I interpret these.

MSeifert04 commented on astropy/io/fits/tests/test_fitsheader.py:
```diff
@@ -0,0 +1,159 @@
+    # Different test because stderr will not be none in case of testing by existing keyword
+    # fitsheader searches for a keyword in all the HDU's
+    @pytest.mark.parametrize('test_kw', [
+        ('RANDOMKEY')
```
This looks like only one test? It doesn't make sense to actually `parametrize` the test, right?

MSeifert04 commented on astropy/io/fits/tests/test_fitsheader.py:
```diff
@@ -0,0 +1,159 @@
+#     with open(request.params[0]) as test_fits:
+#         yield test_fits
+
+@pytest.fixture(scope='module')
```
Is it actually necessary to declare this helper-function as fixture?

mohanagr replied:
Using fixture I wouldn't have to call the function everytime. (lesser executions because `module` scope? lots of method use it).

MSeifert04 replied:
So this is just a way so the regex doesn't need to be compiled multiple times?

mohanagr replied:
Yes. module scope fixtures distribute the return value across functions which need it, don't they? Although I know it's almost no overhead. Should I remove it and replace by functions?

MSeifert04 replied:
Well they seem like a really complicated way (using regex) of testing wheater `str.startswith('WARNING')` or `'WARNING' in str`, right?

mohanagr replied:
I just thought RE might be of use later. I'll change it if required.

pllim replied:
I agree with @MSeifert04 . This seems unnecessarily complicated. Using `startswith()` is more readable.

MSeifert04 commented:
It's my personal opinion but I wouldn't test any expected failures, like specifying the wrong file, non-existing keyword or non-existing extensions.

I think we basically need to make sure the script works and doesn't get accidentally broken. The failures (probably) come from the non-scripty part of the `io.fits` module and should be tested already there.

mohanagr commented:
Doesn't script not emitting warning when it should come under _accidentally broken_ (i.e. bad user inputs and things)?

MSeifert04 commented:
@mohanagr Before you push anything else: I would be very interested to see what travis+coverage reports.

I really need to think a bit more about the scope of the PR and what to test with scripts. In the end some utility that can be used to test scripts would be really nice. Something like you basically did with your `setup` and `teardown` but more dynamic so it can be easily adjusted for other scripts.

Let's wait and see what the others say.

mohanagr commented:
All right. I'd really like to contribute more to this issue and the testing part in general. Guess that'd help me slowly understand the codebase better, so that I can tackle other issues as well.

mohanagr commented:
@MSeifert04 I got your point. The cases for invalid files, extensions etc. has been handled in the _non-script_  part of the script! I'll remove the `xfail` ones in the next commit once Travis finishes builds.

pllim commented on astropy/io/fits/tests/test_fitsheader.py:
```diff
@@ -0,0 +1,161 @@
+
+    @pytest.mark.parametrize('test_kw,expected', [
+        ('BSCALE', 'BSCALE  =           1.000000E0 / REAL = TAPE*BSCALE + BZERO'),
+        pytest.mark.xfail(('LRFWAVE', 'LRFWAVE = 0.0 / linear ramp filter wavelength'), reason = 'bad format')
```
Putting `xfail` inside `parametrize` is hard to read. I prefer `xfail` to be its own thing.

pllim commented on astropy/io/fits/tests/test_fitsheader.py:
```diff
@@ -0,0 +1,161 @@
+class TestFITSheader_script(FitsTestCase):
+
+    def setup_method(self,  method):
+        self.sys_argv_orig = sys.argv
```
Using `subprocess` to spawn the command line would be simpler but I wonder if that would work in Windows; do you know, @MSeifert04 ?

MSeifert04 replied:
There should be no problem with `subprocess`, I've used it several times (mostly to call `zlib`). However, is `coverage` able to deal with subprocesses?

pllim replied:
Good point. Maybe this is not worth pursuing.

mohanagr replied:
@pllim `subprocess` does work on windows AFAIK. Also I will remove the `xfail` things. I understand the usage better now. I will add the test for the table formatting and go on with `fitscheck`.

saimn commented:
It is possible to avoid mocking `sys.argv` by having the `main` function take an optional `args` argument which is passed to `ArgumentParser.parse_args`. This is already possible with the fitsheader script: https://github.com/astropy/astropy/blob/master/astropy/io/fits/scripts/fitsheader.py#L284 (and easy to do for the other scripts).

About the way to test scripts, some options:
- pytest has to `TestDir.run` method to test itself it seems, in a subprocess or not. http://stackoverflow.com/a/13500346/4728722 gives an example.
- there is also https://pypi.python.org/pypi/pytest-console-scripts

Just some thoughts, I'm not sure it's useful to put too much work here.

mohanagr commented:
@saimn @pllim In my opinion `monkeypatch` is the best thing to do. It allows one to change context for a test and then go back. (Like arhv, pwd, syspath etc. Etc.). Simple enough.

mohanagr commented:
@saimn @pllim In my opinion `monkeypatch` is the best thing to do. It allows one to change context for a test and then go back. (Like arhv, pwd, syspath etc. Etc.). Simple enough.

pllim commented:
For the sake of simplicity, I think the current call to `main()` should be okay...

mohanagr commented:
@pllim Sorry I didn't read your comment after pushing. I have taken other comments into account.

- Removed `xfails`
- Fixed need to mock `sys.argv`
I'm deliberately including a few tests where the script should fail by making assertions on output (instead of raising an error myself and using `pytest.raises`) I'll remove those too if necessary.

MSeifert04 commented:
@mohanagr Just out of interest: Did you create tests based on (missing) coverage?

mohanagr commented:
@MSeifert04 Not really. But coverage in this particular file?

mohanagr commented:
Also can you suggest how to test `fitscheck`? I'll manually run that script and compare outputs?

saimn commented:
> @saimn @pllim In my opinion monkeypatch is the best thing to do.

IMO it brings nothing here and only adds useless complexity, as you can use directly `fitsheader.main([...])` and it will work the same.

mohanagr commented:
Can someone please tell me how to go ahead with #5874 ? Should I push the fix here?

pllim commented:
Ideally, the bug fix would be a separate PR and should be merged first, and then this PR to be rebased on top of that once merged. They have to be separate because this PR would go into v2.0 and the bug fix into v1.3.2. Hope this clarifies the matter. I apologize for any inconvenience caused.

mohanagr commented:
@pllim Yes. Thanks. I'll just file a separate PR too. I pushed here so that I could continue testing this. Or I'll to omit one test case. 
P.S. There is really no inconvenience really. Things for me to learn.

pllim commented:
@mohanagr , the affected test can move to the new bug fix PR as well, if possible. This is because usually we ask the bug fix PR to add a test to demonstrate that the bug is indeed fix and to prevent it from creeping back in the future.

saimn commented:
@mohanagr - Are you still interested in this ? If yes, to make progress, I would suggest to focus here on `test_fitsheader.py`, and to limit the number of things to test.

mohanagr commented:
Yes, I'll continue this and try to fix the test scripts.

On May 15, 2017 3:43 PM, "Simon Conseil" <notifications@github.com> wrote:

@mohanagr <https://github.com/mohanagr> - Are you still interested in this
? If yes, to make progress, I would suggest to focus here on
test_fitsheader.py, and to limit the number of things to test.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/astropy/astropy/pull/5858#issuecomment-301434142>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AOc7f9pA6cCfWYWUvBbn0JOjiqi3EN7uks5r6CVIgaJpZM4MSdgI>
.

saimn commented on astropy/io/fits/tests/test_fitsheader.py:
```diff
@@ -0,0 +1,122 @@
+from ..hdu import PrimaryHDU
+from ..scripts import fitsheader
+from ....tests.helper import catch_warnings
+from ....tests.helper import pytest
```
Could you rebase on master, and import pytest directly (the bundle version was removed recently).

saimn commented on astropy/io/fits/tests/test_fitsheader.py:
```diff
@@ -0,0 +1,122 @@
+import re
+
+from . import FitsTestCase
+from ..hdu import PrimaryHDU
```
Please remove the imports that are not used (`PrimaryHDU`, `AstropyDeprecationWarning`).

saimn commented:
@mohanagr - Ok, great. So I think you should start with testing the normal usage of the script: choose a test file and simply test that the output is the expected one (with or without `-e`). 
With a test file whose header is not too long (I would choose `astropy/io/fits/tests/data/table.fits`) you can directly compare the output with the expected one. The goal is really to test the script, not the behavior of `fits.Header` which is already well tested. So there is no real need for all the pytest magic here (test one keyword with `-k` and remove the `parametrize` uses, remove `monkeypatch` as explained above), and keep the errors testing and edge cases for the end.

pllim commented:
This was created during GSoC 2017 application period, which is now over. @mohanagr , if you are interested to finish this, feel free to re-open, but I am closing this for now due to inactivity.
